File: src/pythonnative/native/android/src/main/java/com/pythonnative/runtime/assets/PNAssets.kt
package com.pythonnative.runtime.assets

import android.content.Context
import android.content.res.AssetManager
import android.graphics.Typeface
import com.pythonnative.generated.PNAssetManifest
import com.pythonnative.generated.PNFontFace
import com.pythonnative.runtime.bridge.MainThread
import com.pythonnative.runtime.bridge.PNLog
import org.json.JSONObject
import java.io.File
import java.io.FileInputStream
import java.io.FileNotFoundException
import java.io.InputStream
import kotlin.math.abs

/**
 * Resolves `asset://` URIs against the bundled `app/assets/` folder of
 * the APK and, while connected to `pn start`, the dev overlay that
 * shadows it.
 *
 * `pn build` writes a `pn_assets.json` manifest next to the staged
 * assets listing every file, the density variants of each base name
 * (`logo@2x.png`), and the parsed font faces. The overlay manifest
 * arrives through the `Assets` module. Resolution looks in the overlay
 * first, then the bundle, and picks the variant closest to the screen
 * density (preferring the next denser one so images are downsampled,
 * not upsampled).
 */
object PNAssets {
    /** The URI scheme Python emits for bundled files. */
    const val SCHEME = "asset://"
    /** Where the builder stages `app/assets/` inside the APK's assets. */
    const val BUNDLE_ROOT = "app/assets"

    /** One resolved file: how to open it and the density it was drawn for. */
    class Resolved(
        val path: String,
        val scale: Float,
        /** Overlay files live on disk; bundled ones only inside the APK. */
        val file: File?,
        private val opener: () -> InputStream,
    ) {
        fun open(): InputStream = opener()
    }

    private class Root(val manifest: PNAssetManifest?, val directory: File?, val assets: AssetManager?) {
        fun exists(path: String): Boolean {
            if (directory != null) return File(directory, path).isFile
            val manager = assets ?: return false
            return try { manager.open("$BUNDLE_ROOT/$path").close(); true } catch (_: Exception) { false }
        }
        fun open(path: String): InputStream {
            if (directory != null) return FileInputStream(File(directory, path))
            return assets?.open("$BUNDLE_ROOT/$path") ?: throw FileNotFoundException(path)
        }
        fun resolved(path: String): Resolved =
            Resolved(path, scaleOf(path), directory?.let { File(it, path) }) { open(path) }
    }

    private val lock = Any()
    private var overlay: Root? = null
    private var bundled: Root? = null
    private var bundledContext: Context? = null
    private val typefaces = HashMap<String, Typeface?>()
    private var faces: List<Pair<PNFontFace, Root>> = emptyList()
    private val listeners = ArrayList<() -> Unit>()
    @Volatile var generation = 0
        private set

    /** Attach the app context so bundled assets can be opened; idempotent. */
    fun attach(context: Context) {
        synchronized(lock) {
            if (bundledContext === context.applicationContext && bundled != null) return
            bundledContext = context.applicationContext
            val assets = context.applicationContext.assets
            bundled = Root(readManifest { assets.open("$BUNDLE_ROOT/pn_assets.json") }, null, assets)
            typefaces.clear()
            rebuildFaces()
        }
    }

    /** Point the resolver at a dev overlay (or clear it with `null`) and notify listeners. */
    fun configure(overlay: String?, manifest: PNAssetManifest?) {
        synchronized(lock) {
            this.overlay = if (overlay.isNullOrEmpty()) null else Root(manifest, File(overlay), null)
            typefaces.clear()
            rebuildFaces()
            generation++
        }
        val snapshot = synchronized(lock) { listeners.toList() }
        MainThread.runOnMain { for (listener in snapshot) listener() }
    }

    /** Register a callback for overlay changes; returns the unsubscribe action. */
    fun addListener(listener: () -> Unit): () -> Unit {
        synchronized(lock) { listeners.add(listener) }
        return { synchronized(lock) { listeners.remove(listener) } }
    }

    /** Whether `value` is an `asset://` URI. */
    fun isAssetUri(value: String?): Boolean = value?.startsWith(SCHEME) == true

    /** The asset path named by a URI (`asset://images/a.png` -> `images/a.png`). */
    fun pathOf(uri: String): String = uri.removePrefix(SCHEME).trimStart('/')

    /** Resolve a path or URI, preferring the overlay and the best density variant. */
    fun resolve(pathOrUri: String, scale: Float = deviceScale()): Resolved? {
        val path = pathOf(pathOrUri)
        for (root in roots()) {
            val manifest = root.manifest
            if (manifest != null) {
                val variant = variant(manifest, path, scale)
                if (variant != null && root.exists(variant)) return root.resolved(variant)
            }
            if (root.exists(path)) return root.resolved(path)
        }
        return null
    }

    /** Whether some copy of `pathOrUri` exists. */
    fun exists(pathOrUri: String): Boolean = resolve(pathOrUri) != null

    /** Read the bytes of an asset (the base file, not a density variant). */
    fun read(pathOrUri: String): ByteArray {
        val path = pathOf(pathOrUri)
        for (root in roots()) {
            if (root.exists(path)) return root.open(path).use { it.readBytes() }
        }
        throw FileNotFoundException("asset not found: $path")
    }

    /** Pick the manifest variant of `path` for `scale`, or null when the manifest has no entry. */
    fun variant(manifest: PNAssetManifest, path: String, scale: Float): String? {
        val base = baseName(path)
        val group = manifest.variants[base]
        if (!group.isNullOrEmpty()) {
            val options = group.mapNotNull { (key, value) -> key.toFloatOrNull()?.let { it to value } }
            options.firstOrNull { abs(it.first - scale) < 0.001f }?.let { return it.second }
            options.filter { it.first > scale }.minByOrNull { it.first }?.let { return it.second }
            return options.maxByOrNull { it.first }?.second
        }
        return if (manifest.files.contains(base)) base else null
    }

    /** `images/logo@2x.png` -> `images/logo.png`. */
    fun baseName(path: String): String {
        val slash = path.lastIndexOf('/')
        val directory = if (slash >= 0) path.substring(0, slash + 1) else ""
        val name = path.substring(slash + 1)
        val dot = name.lastIndexOf('.')
        val stem = if (dot > 0) name.substring(0, dot) else name
        val ext = if (dot > 0) name.substring(dot) else ""
        val at = stem.lastIndexOf('@')
        if (at >= 0 && stem.endsWith("x") && stem.substring(at + 1, stem.length - 1).toFloatOrNull() != null) {
            return directory + stem.substring(0, at) + ext
        }
        return path
    }

    /** The density encoded in a variant name (`@2x` -> 2), or 1. */
    fun scaleOf(path: String): Float {
        val name = path.substringAfterLast('/')
        val stem = name.substringBeforeLast('.', name)
        val at = stem.lastIndexOf('@')
        if (at < 0 || !stem.endsWith("x")) return 1f
        return stem.substring(at + 1, stem.length - 1).toFloatOrNull() ?: 1f
    }

    /** The screen density as an iOS-style scale (`xxhdpi` -> 3). */
    fun deviceScale(): Float {
        val context = bundledContext ?: return 1f
        val density = context.resources.displayMetrics.density
        return if (density > 0f) density else 1f
    }

    // ------------------------------------------------------------------
    // Fonts
    // ------------------------------------------------------------------

    /** Every bundled font face (overlay faces shadow bundle faces of the same family/weight/style). */
    fun fontFaces(): List<PNFontFace> = synchronized(lock) { faces.map { it.first } }

    /** The best face for `family`, or null when the family isn't bundled. */
    fun fontFace(family: String, weight: Int, italic: Boolean): PNFontFace? {
        val wanted = family.trim().lowercase()
        val candidates = synchronized(lock) { faces.filter { it.first.family.lowercase() == wanted } }
        if (candidates.isEmpty()) return null
        return candidates.minByOrNull { (face, _) ->
            (if (face.italic == italic) 0 else 1000) + abs(face.weight.toInt() - weight)
        }?.first
    }

    /**
     * A `Typeface` for the bundled face closest to `family`/`weight`/`italic`,
     * or null when the family isn't bundled (callers fall back to system fonts).
     */
    fun typeface(family: String, weight: Int, italic: Boolean): Typeface? {
        val face = fontFace(family, weight, italic) ?: return null
        val key = "${face.family.lowercase()}|${face.weight}|${face.italic}"
        synchronized(lock) {
            if (typefaces.containsKey(key)) return typefaces[key]
        }
        val root = synchronized(lock) { faces.firstOrNull { it.first === face }?.second }
        val loaded = try {
            when {
                root == null -> null
                root.directory != null -> Typeface.createFromFile(File(root.directory, face.path))
                root.assets != null -> Typeface.createFromAsset(root.assets, "$BUNDLE_ROOT/${face.path}")
                else -> null
            }
        } catch (error: Exception) {
            PNLog.rateLimited("font:${face.path}", "[assets] could not load ${face.path}", error)
            null
        }
        synchronized(lock) { typefaces[key] = loaded }
        return loaded
    }

    private fun rebuildFaces() {
        val merged = ArrayList<Pair<PNFontFace, Root>>()
        val seen = HashSet<String>()
        for (root in listOfNotNull(overlay, bundled)) {
            for (face in root.manifest?.fonts ?: emptyList()) {
                val key = "${face.family.lowercase()}|${face.weight}|${face.italic}"
                if (!seen.add(key)) continue
                merged.add(face to root)
            }
        }
        faces = merged
    }

    // ------------------------------------------------------------------
    // Internals
    // ------------------------------------------------------------------

    private fun roots(): List<Root> = synchronized(lock) { listOfNotNull(overlay, bundled) }

    private fun readManifest(open: () -> InputStream): PNAssetManifest? = try {
        val text = open().use { it.readBytes().toString(Charsets.UTF_8) }
        PNAssetManifest.decode(JSONObject(text))
    } catch (_: Exception) {
        null
    }

    /** Testing hook: install a directory as the bundle root instead of the APK assets. */
    internal fun installBundleForTesting(directory: File?, manifest: PNAssetManifest?) {
        synchronized(lock) {
            bundled = if (directory == null) null else Root(manifest, directory, null)
            overlay = null
            typefaces.clear()
            rebuildFaces()
        }
    }
}

File: src/pythonnative/native/ios/Sources/PythonNativeKit/Support/PNAssets.swift
import CoreText
import UIKit

/// Resolves `asset://` URIs against the bundled `app/assets/` directory
/// and, while connected to `pn start`, the dev overlay that shadows it.
///
/// The bundle ships a `pn_assets.json` manifest written by `pn build`
/// that lists every file, groups density variants (`logo@2x.png`), and
/// carries the parsed font faces. The overlay manifest arrives over the
/// bridge through the `Assets` module. Resolution looks in the overlay
/// first, then the bundle, and picks the variant closest to the screen
/// scale (preferring the next denser one so images are downsampled, not
/// upsampled).
public final class PNAssets {
    public static let shared = PNAssets()

    /// The URI scheme Python emits for bundled files.
    public static let scheme = "asset://"
    /// Posted on the main queue after `configure` swaps the overlay.
    public static let didChange = Notification.Name("PNAssetsDidChange")

    /// One resolved file: where it lives and the density it was drawn for.
    public struct Resolved: Equatable {
        public let url: URL
        public let scale: CGFloat
        public let path: String
    }

    private struct Root {
        let directory: URL
        let manifest: PNAssetManifest?
    }

    private let lock = NSLock()
    private var overlay: Root?
    private var bundled: Root?
    private var registeredFonts: Set<String> = []
    private var faces: [PNFontFace] = []
    private(set) public var generation = 0

    init(bundleRoot: URL? = nil) {
        let root = bundleRoot ?? Bundle.main.resourceURL?.appendingPathComponent("app/assets", isDirectory: true)
        if let root = root, FileManager.default.fileExists(atPath: root.path) {
            bundled = Root(directory: root, manifest: Self.readManifest(root))
        }
        rebuildFonts()
    }

    // MARK: - Configuration

    /// Point the resolver at a dev overlay (or clear it with `nil`).
    public func configure(overlay directory: String?, manifest: PNAssetManifest?) {
        lock.lock()
        if let directory = directory, !directory.isEmpty {
            overlay = Root(directory: URL(fileURLWithPath: directory, isDirectory: true), manifest: manifest)
        } else {
            overlay = nil
        }
        generation += 1
        lock.unlock()
        rebuildFonts()
        PNMain.run { NotificationCenter.default.post(name: PNAssets.didChange, object: self) }
    }

    /// Whether `value` is an `asset://` URI.
    public static func isAssetURI(_ value: String?) -> Bool {
        value?.hasPrefix(scheme) == true
    }

    /// The asset path named by a URI (`asset://images/a.png` -> `images/a.png`).
    public static func path(of uri: String) -> String {
        var text = uri
        if text.hasPrefix(scheme) { text.removeFirst(scheme.count) }
        while text.hasPrefix("/") { text.removeFirst() }
        return text
    }

    // MARK: - Resolution

    /// Resolve a path or URI to a file, preferring the overlay and the best density variant.
    public func resolve(_ pathOrURI: String, scale: CGFloat = PNWindow.screenScale()) -> Resolved? {
        let path = Self.path(of: pathOrURI)
        lock.lock()
        let roots = [overlay, bundled].compactMap { $0 }
        lock.unlock()
        for root in roots {
            if let manifest = root.manifest, let variant = Self.variant(in: manifest, path: path, scale: scale) {
                let url = root.directory.appendingPathComponent(variant)
                if FileManager.default.fileExists(atPath: url.path) {
                    return Resolved(url: url, scale: Self.scale(of: variant), path: variant)
                }
            }
            let url = root.directory.appendingPathComponent(path)
            if FileManager.default.fileExists(atPath: url.path) {
                return Resolved(url: url, scale: Self.scale(of: path), path: path)
            }
        }
        return nil
    }

    /// Whether some copy of `pathOrURI` exists.
    public func exists(_ pathOrURI: String) -> Bool {
        resolve(pathOrURI) != nil
    }

    /// Read the bytes of an asset (the base file, not a density variant).
    public func read(_ pathOrURI: String) throws -> Data {
        let path = Self.path(of: pathOrURI)
        lock.lock()
        let roots = [overlay, bundled].compactMap { $0 }
        lock.unlock()
        for root in roots {
            let url = root.directory.appendingPathComponent(path)
            if FileManager.default.fileExists(atPath: url.path) { return try Data(contentsOf: url) }
        }
        throw PNAssetError.notFound(path)
    }

    /// Pick the manifest variant of `path` for `scale`, or nil when the manifest has no entry.
    static func variant(in manifest: PNAssetManifest, path: String, scale: CGFloat) -> String? {
        let base = baseName(path)
        if let group = manifest.variants[base], !group.isEmpty {
            var options: [(CGFloat, String)] = []
            for (key, value) in group {
                if let number = Double(key) { options.append((CGFloat(number), value)) }
            }
            if let exact = options.first(where: { abs($0.0 - scale) < 0.001 }) { return exact.1 }
            let above = options.filter { $0.0 > scale }.sorted { $0.0 < $1.0 }
            if let next = above.first { return next.1 }
            return options.max { $0.0 < $1.0 }?.1
        }
        return manifest.files.contains(base) ? base : nil
    }

    /// `images/logo@2x.png` -> `images/logo.png`.
    static func baseName(_ path: String) -> String {
        let url = URL(fileURLWithPath: path)
        let ext = url.pathExtension
        var stem = url.deletingPathExtension().lastPathComponent
        if let at = stem.lastIndex(of: "@"), stem.hasSuffix("x"), Double(stem[stem.index(after: at)..<stem.index(before: stem.endIndex)]) != nil {
            stem = String(stem[..<at])
        }
        let directory = (path as NSString).deletingLastPathComponent
        let name = ext.isEmpty ? stem : "\(stem).\(ext)"
        return directory.isEmpty ? name : "\(directory)/\(name)"
    }

    /// The density encoded in a variant name (`@2x` -> 2), or 1.
    static func scale(of path: String) -> CGFloat {
        let stem = URL(fileURLWithPath: path).deletingPathExtension().lastPathComponent
        guard let at = stem.lastIndex(of: "@"), stem.hasSuffix("x"),
              let value = Double(stem[stem.index(after: at)..<stem.index(before: stem.endIndex)]) else { return 1 }
        return CGFloat(value)
    }

    // MARK: - Fonts

    /// Every bundled font face (overlay faces shadow bundle faces of the same family/weight/style).
    public var fontFaces: [PNFontFace] {
        lock.lock()
        defer { lock.unlock() }
        return faces
    }

    /// The PostScript name of the best face for `family`, or nil when the family isn't bundled.
    public func fontName(family: String, weight: Int, italic: Bool) -> String? {
        let wanted = family.trimmingCharacters(in: .whitespaces).lowercased()
        let candidates = fontFaces.filter { $0.family.lowercased() == wanted }
        guard !candidates.isEmpty else { return nil }
        let best = candidates.min { lhs, rhs in
            let l = (lhs.italic == italic ? 0 : 1000) + abs(Int(lhs.weight) - weight)
            let r = (rhs.italic == italic ? 0 : 1000) + abs(Int(rhs.weight) - weight)
            return l < r
        }
        return best?.postscript_name
    }

    private func rebuildFonts() {
        lock.lock()
        let roots = [overlay, bundled].compactMap { $0 }
        lock.unlock()
        var merged: [PNFontFace] = []
        var seen: Set<String> = []
        for root in roots {
            for face in root.manifest?.fonts ?? [] {
                let key = "\(face.family.lowercased())|\(face.weight)|\(face.italic)"
                if seen.contains(key) { continue }
                seen.insert(key)
                let url = root.directory.appendingPathComponent(face.path)
                register(url)
                merged.append(face)
            }
        }
        lock.lock()
        faces = merged
        lock.unlock()
    }

    private func register(_ url: URL) {
        lock.lock()
        let already = registeredFonts.contains(url.path)
        if !already { registeredFonts.insert(url.path) }
        lock.unlock()
        if already { return }
        var error: Unmanaged<CFError>?
        if !CTFontManagerRegisterFontsForURL(url as CFURL, .process, &error) {
            let description = (error?.takeRetainedValue()).map { CFErrorCopyDescription($0) as String } ?? "unknown error"
            // Already-registered fonts (bundle plus overlay) report an error that isn't a problem.
            if !description.contains("already") {
                PNLog.rateLimited(PNLog.components, key: "font:\(url.lastPathComponent)", "[assets] could not register \(url.lastPathComponent): \(description)")
            }
        }
    }

    // MARK: - Manifest

    static func readManifest(_ directory: URL) -> PNAssetManifest? {
        let url = directory.appendingPathComponent("pn_assets.json")
        guard let data = try? Data(contentsOf: url),
              let object = try? JSONSerialization.jsonObject(with: data) else { return nil }
        return try? PNValues.decode(PNAssetManifest.self, object)
    }
}

public enum PNAssetError: Error, LocalizedError {
    case notFound(String)

    public var errorDescription: String? {
        switch self {
        case .notFound(let path): return "asset not found: \(path)"
        }
    }
}

/// `Assets`: the bridge face of `PNAssets` (overlay configuration and raw reads).
public final class AssetsModule: AssetsImplementation {
    public init() {}

    public func configure(overlay: String?, manifest: PNAssetManifest) throws {
        PNAssets.shared.configure(overlay: overlay, manifest: manifest)
    }

    public func exists(path: String) throws -> Bool {
        PNAssets.shared.exists(path)
    }

    public func read(path: String) throws -> String? {
        try PNAssets.shared.read(path).base64EncodedString()
    }
}
